ctonly/ct.go

```diff
@@ -0,0 +1,169 @@
+package ctonly
+
+import (
+	"crypto/sha256"
+	"errors"
+
+	"github.com/transparency-dev/merkle/rfc6962"
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// Entry represents a CT log entry.
+type Entry struct {
+	Timestamp          uint64
+	IsPrecert          bool
+	Certificate        []byte
+	Precertificate     []byte
+	PrecertSigningCert []byte
+	IssuerKeyHash      []byte
+}
+
+// LeafData returns the data which should be added to an entry bundle for this entry.
+//
+// Note that this will include data which IS NOT directly committed to by the entry's
+// MerkleLeafHash.
+func (c Entry) LeafData(idx uint64) []byte {
+	b := cryptobyte.NewBuilder([]byte{})
+	b.AddUint64(uint64(c.Timestamp))
+	if !c.IsPrecert {
+		b.AddUint16(0 /* entry_type = x509_entry */)
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.Certificate)
+		})
+	} else {
+		b.AddUint16(1 /* entry_type = precert_entry */)
+		b.AddBytes(c.IssuerKeyHash[:])
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.Certificate)
+		})
+	}
+	addExtensions(b, idx)
+	if c.IsPrecert {
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.Precertificate)
+		})
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.PrecertSigningCert)
+		})
+	}
+	return b.BytesOrPanic()
+}
+
+// MerkleLeafHash returns the RFC6962 leaf hash for this entry.
+//
+// Note that we embed an SCT extension which captures the index of the entry in the log according to
+// the mechanism specified in https://c2sp.org/ct-static-api.
+func (c Entry) MerkleLeafHash(leafIndex uint64) []byte {
+	b := &cryptobyte.Builder{}
+	b.AddUint8(0 /* version = v1 */)
+	b.AddUint8(0 /* leaf_type = timestamped_entry */)
+	b.AddUint64(uint64(c.Timestamp))
+	if !c.IsPrecert {
+		b.AddUint16(0 /* entry_type = x509_entry */)
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.Certificate)
+		})
+	} else {
+		b.AddUint16(1 /* entry_type = precert_entry */)
+		b.AddBytes(c.IssuerKeyHash[:])
+		b.AddUint24LengthPrefixed(func(b *cryptobyte.Builder) {
+			b.AddBytes(c.Certificate)
+		})
+	}
+	addExtensions(b, leafIndex)
+	return rfc6962.DefaultHasher.HashLeaf(b.BytesOrPanic())
+}
+
+func (c Entry) Identity() []byte {
+	var r [sha256.Size]byte
+	if c.IsPrecert {
+		r = sha256.Sum256(c.Precertificate)
+	} else {
+		r = sha256.Sum256(c.Certificate)
+	}
+	return r[:]
+}
+
+func addExtensions(b *cryptobyte.Builder, leafIndex uint64) {
+	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+		ext, err := extensions{LeafIndex: leafIndex}.Marshal()
+		if err != nil {
+			b.SetError(err)
+			return
+		}
+		b.AddBytes(ext)
+	})
+}
+
+// extensions is the CTExtensions field of SignedCertificateTimestamp and
+// TimestampedEntry, according to c2sp.org/static-ct-api.
+type extensions struct {
+	LeafIndex uint64
+}
+
+func (c extensions) Marshal() ([]byte, error) {
+	// enum {
```

Comment:
Are these old notes, or a reference to a different doc? If they're references, please link to it.

Reply:
This code is cut'n'pasted from the sunlight implementation - that's linked above in the comments at the top of the file (Lines #16 and #35).